include/verilated_vpi.cpp

```diff
@@ -2812,13 +2824,1086 @@ vpiHandle vpi_put_value(vpiHandle object, p_vpi_value valuep, p_vpi_time /*time_
 }
 
-void vpi_get_value_array(vpiHandle /*object*/, p_vpi_arrayvalue /*arrayvalue_p*/,
-                         PLI_INT32* /*index_p*/, PLI_UINT32 /*num*/) {
-    VL_VPI_UNIMP_();
+bool vl_check_array_format(const VerilatedVar* varp, const p_vpi_arrayvalue arrayvalue_p,
+                            const char* fullname) {
+    bool status = true;
+
+    if (arrayvalue_p->format == vpiVectorVal) {
+        switch (varp->vltype()) {
+            case VLVT_UINT8:
+            case VLVT_UINT16:
+            case VLVT_UINT32:
+            case VLVT_UINT64:
+            case VLVT_WDATA:
+                return status;
+            default:
+                status = false;
+        }
+    } else if (arrayvalue_p->format == vpiIntVal) {
+        switch (varp->vltype()) {
+            case VLVT_UINT8:
+            case VLVT_UINT16:
+            case VLVT_UINT32:
+                return status;
+            default:
+                status = false;
+        }
+    } else if (arrayvalue_p->format == vpiRealVal) {
+        switch (varp->vltype()) {
+            case VLVT_UINT64:
+                return status;
+            default:
+                status = false;
+        }
+    } else if ((arrayvalue_p->format == vpiRawTwoStateVal)
+        || (arrayvalue_p->format == vpiRawFourStateVal)) {
+        switch (varp->vltype()) {
+            case VLVT_UINT8:
+            case VLVT_UINT16:
+            case VLVT_UINT32:
+            case VLVT_UINT64:
+            case VLVT_WDATA:
+                return status;
+            default:
+                status = false;
+        }
+    } else if (arrayvalue_p->format == vpiShortIntVal) {
+        switch (varp->vltype()) {
+            case VLVT_UINT8:
+            case VLVT_UINT16:
+                return status;
+            default:
+                status = false;
+        }
+    } else if (arrayvalue_p->format == vpiLongIntVal) {
+        switch (varp->vltype()) {
+            case VLVT_UINT8:
+            case VLVT_UINT16:
+            case VLVT_UINT64:
+                return status;
+            default:
+                status = false;
+        }
+    }
+
+    VL_VPI_ERROR_(__FILE__, __LINE__, "%s: Unsupported format (%s) as requested for %s",
+                __func__, VerilatedVpiError::strFromVpiVal(arrayvalue_p->format), fullname);
+
+    return status;
 }
-void vpi_put_value_array(vpiHandle /*object*/, p_vpi_arrayvalue /*arrayvalue_p*/,
-                         PLI_INT32* /*index_p*/, PLI_UINT32 /*num*/) {
-    VL_VPI_UNIMP_();
+
+void vl_get_value_array(vpiHandle object, p_vpi_arrayvalue arrayvalue_p,
+                        PLI_INT32* index_p, PLI_UINT32 num) {
+    const VerilatedVpioVar* const vop = VerilatedVpioVar::castp(object);
+    if (!vl_check_array_format(vop->varp(), arrayvalue_p, vop->fullname())) return;
+
+    const VerilatedVar* const varp = vop->varp();
+    int size = vop->size();
+    int index = index_p[0];
+
+    static thread_local EData out_ptr[VL_VALUE_STRING_MAX_WORDS * 2];
```

Comment:
Doesn't the user provide the data storage?

Also this isn't a pointer, so should be just named e.g. out_data.

Reply:
Thank you for all the detailed feedback. I'll be fixing everything ASAP.

For this, the user can provide the data storage, signalled by the vpiUserAllocFlag. If not present, the default behaviour is to provide memory.

> By default, array values shall be returned in memory allocated by VPI [...] However, if the application sets the vpiUserAllocFlag in the value.flags field, the function will assume the calling application has set the value field to point to a buffer of sufficient size allocated by the application for placing the values.

Reply:
>However, if the application sets the vpiUserAllocFlag in the value.flags field, the function will assume the calling application has set the value field to point to a buffer of sufficient size allocated by the application for placing the values.

I think we should just consider that unsupported behavior, do you really require it?  It's nasty, and has a lot of edge cases we need to check.

Reply:
It's not really required, it would only limit the amount of words returned to VL_VALUE_STRING_MAX_WORDS.
We can completely leave vpiUserAllocFlag unsupported, with no issue.